src/clipping/weileratherton.jl

```diff
@@ -0,0 +1,408 @@
+
+  function RingVertex{VT,M,C}(point::Point{M,C}) where {VT<:VertexType,M<:Manifold,C<:CRS}
+    v = new(point)
+    v.left = v
+    v.right = v
+    return v
+  end
+end
+
+RingVertex{VT}(point::Point{M,C}) where {VT<:VertexType,M<:Manifold,C<:CRS} =
+  RingVertex{VT,manifold(point),typeof(point.coords)}(point)
+
+function appendvertices!(v1::RingVertex{Normal}, v2::RingVertex{Normal})
+  v2.left = v1.left
+  v1.left = v2
+  v2.right = v1.right
+  v1.right = v2
+end
+
+function gettraversalring(ring::Ring)
+  vs = vertices(ring)
+  start = RingVertex{Normal}(vs[1])
+
+  for v in vs[2:end]
+    new = RingVertex{Normal}(v)
+    appendvertices!(start, new)
+    start = new
+  end
+
+  return start.left
+end
+
+isnormal(_::RingVertex{Normal}) = true
+isnormal(_::RingVertex) = false
+
+function insertintersection!(head::RingVertex, intersection::RingVertex, side::Symbol)
+  tail = head.left
+  vertex = head
+
+  new = measure(Segment(head.point, intersection.point))
+  while true
+    os = isnormal(vertex) ? :right : side
+    current = measure(Segment(head.point, getfield(vertex, os).point))
+    if (new < current) || (getfield(vertex, os) == tail)
+      next = getfield(vertex, os)
+      setfield!(intersection, side, next)
+      setfield!(vertex, os, intersection)
+      break
+    end
+    vertex = getfield(vertex, os)
+  end
+end
+
+# Traversing and selecting following vertices for collecting of the rings after clipping depends
+# on the vertex type. Helper nextvertex follows the correct path.
+nextvertex(v::RingVertex{Normal}) = v.right
+nextvertex(v::RingVertex{Entering}) = v.right
+nextvertex(v::RingVertex{Exiting}) = v.left
+
+function collectclipped(entering::Vector{RingVertex{Entering}})
+  rings::Vector{Ring} = []
+  visited::Vector{RingVertex} = []
+  for i in eachindex(entering)
+    if entering[i] in visited
+      continue
+    end
+
+    ring::Vector{RingVertex} = []
+    vertex = entering[i]
+    while !(vertex in ring)
+      if vertex in visited
+        break
+      end
+
+      push!(ring, vertex)
+      push!(visited, vertex)
+      vertex = nextvertex(vertex)
+    end
+
+    # Remove duplicates.
+    newring::Vector{RingVertex} = [ring[1]]
+    for i in 2:length(ring)
+      if !isapprox(ring[i].point, newring[end].point)
+        push!(newring, ring[i])
+      end
+    end
+    ring = newring
+
+    # Polygon might start several vertices after the first collected.
+    # This generally happens when there are overlapping edges that lead
+    # to several entering vertices without exiting in between. Then, the
+    # actual polygon is found by discarding the extra vertices before the
+    # proper loop.
+    k = findfirst(x -> ring[end].point == x.point, ring[1:(end - 1)])
+    if !isnothing(k)
+      ring = ring[(k + 1):end]
+    end
+
+    if length(ring) > 2
+      push!(rings, Ring([r.point for r in ring]...))
+    end
+  end
+  return rings
+end
+
+function decide_direction(clipped₁, clipped₂, clipping₁, clipping₂)
```

Comment:
appreciate if you can explain the main goal of these auxiliary functions so that we can reuse existing functionality. for example, we have the `\angle` between rays and points which should be used instead of explicit calls to `atan`, see angles.jl

Reply:
I hope the explanations are sufficient. Changed the explicit usage of atan to use the \angle instead. I still had to convert from -pi:pi to 0:2pi range. One thing that surfaced is that the precision was lower than with my original atan solution, which in turn showed that the isapprox is not properly using atol with radians. So I had to handle atols explicitly.

Also, this change shown that zero length segments are being passed to the function, which then used them incorrectly to somehow get the correct results. Now this is circumvented outside code where it is ensured that in the cases where it matters, the zero  length segments are not passed. Still, sometimes zero length segments are passed in situations where it should not matter and then the function just returns nothing, signaling to the caller to drop the vertex. Maybe this should be also handled outside, but this way it is only handled in one place.